### PR TITLE
[io] Validate ZSTD target/source sizes and compr level

### LINKED ISSUE #9334: Lack of size validation in ZSTD compression
- [ ] Checked for duplicates

### Describe the bug
The size validation is missing in https://github.com/root-project/root/blob/master/core/zstd/src/ZipZSTD.cxx line 47.
This leads to invalid stream in case of sizes between 0x1000000 and 0xffffffff as only the 3 last bytes of the length are put in the encoded data.

### Expected behavior
The size should be checked as done in https://github.com/root-project/root/blob/master/core/lzma/src/ZipLZMA.c line 40 to 42

### To Reproduce
Just try to deflate something of size > 0xffffff

### PR DESCRIPTION
**Pull request by dpiparo**
This PR fixes #9334

### CODE REVIEW

**Comment by dpiparo** [comment]
I found line https://github.com/root-project/root/pull/15038/files#diff-c35e46fbb246941a30fd3fa261506e5eb03b601e3bf6734460f814fd3aa156c9R46 intriguing. We are doubling the compression level given in input. The ZSTD documentation does not say anything which could possibly justify this behaviour https://facebook.github.io/zstd/zstd_manual.html . This deserves a bit of caution/discussion I believe, also given the test of the compression level which has to be coherent with that.

File: core/zstd/src/ZipZSTD.cxx [hunk 1]
```diff
@@ -24,11 +24,22 @@ static const size_t errorCodeSmallBuffer = (size_t)-70;
 
 void R__zipZSTD(int cxlevel, int *srcsize, char *src, int *tgtsize, char *tgt, int *irep)
 {
+
+    if (*tgtsize <= 0) {
```
Comment (ferdymercury) [inline core/zstd/src/ZipZSTD.cxx:28]:
```suggestion
    if (*tgtsize <= kHeaderSize) {
```

Comment (hahnjo) [reply core/zstd/src/ZipZSTD.cxx:28]: This is validated at a higher level:
https://github.com/root-project/root/blob/239cafe11e41a862c35e1f9d0e120cc2a73a3654/core/zip/src/RZip.cxx#L87-L90
(introduced by https://github.com/root-project/root/commit/23261a631c9e53a2ca7981a61d174c6465873a0e; in general, there was some work in this area in the context of https://github.com/root-project/root/pull/14523)

**Comment by hahnjo** [comment]
> I found line https://github.com/root-project/root/pull/15038/files#diff-c35e46fbb246941a30fd3fa261506e5eb03b601e3bf6734460f814fd3aa156c9R46 intriguing. We are doubling the compression level given in input. The ZSTD documentation does not say anything which could possibly justify this behaviour https://facebook.github.io/zstd/zstd_manual.html . This deserves a bit of caution/discussion I believe, also given the test of the compression level which has to be coherent with that.

From the linked manual:

> The library supports regular compression levels from 1 up to ZSTD_maxCLevel(),
> which is currently 22. Levels >= 20, labeled `--ultra`, should be used with
> caution, as they require more memory. The library also offers negative
> compression levels, which extend the range of speed vs. ratio preferences.
> The lower the level, the faster the speed (at the cost of compression).

So the doubling maps ROOT's single digit layer to the "reasonable" level 0 to 18.

**Comment by hahnjo** [comment]
As mentioned inline, the check of `tgtsize` is not needed since https://github.com/root-project/root/commit/23261a631c9e53a2ca7981a61d174c6465873a0e. The check for `*srcsize > 0xffffff` should likely be moved to the upper layer *with proper error reporting* (not just silently `return`ing).

**Pull request closed without merging**